### PR TITLE
Enabled the static analysis of java files.

### PR DESCRIPTION
**Pull request by ReapOmen**
This introduces `infer` into the static-analysis bot. When a `.java` file is part of a revision, it is passed to the `./mach static-analysis check-java` command.

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -26,3 +26,4 @@
 /src/**/yarn-error.log
 /tmp
 /w
+*~
```
Comment (La0) [inline .gitignore]: Could you be a bit more specific here ? Add a folder ? That seems a bit broad

File: nix/gecko_env.nix [hunk 2]
```diff
@@ -60,8 +62,9 @@ in gecko.overrideDerivation (old: {
 
     # Build LDFLAGS and LIBRARY_PATH
     echo "export LDFLAGS=\"$NIX_LDFLAGS\"" >> $geckoenv
-    echo "export LIBRARY_PATH=\"$CMAKE_LIBRARY_PATH\"" >> $geckoenv
-    echo "export LD_LIBRARY_PATH=\"$CMAKE_LIBRARY_PATH\"" >> $geckoenv
+    echo "export SQLITE_LIB=${sqlite.out}/lib/" >> $geckoenv
```
Comment (La0) [inline nix/gecko_env.nix]: Is this environment variable really used by sqlite ?

File: nix/gecko_env.nix [hunk 3]
```diff
@@ -78,6 +81,16 @@ in gecko.overrideDerivation (old: {
     ac_add_options --with-clang-path=${clang_4}/bin/clang
     ac_add_options --with-libclang-path=${llvmPackages_4.libclang}/lib
     mk_add_options AUTOCLOBBER=1
+
+    # Build Firefox for Android:
+    ac_add_options --enable-application=mobile/android
+    ac_add_options --target=arm-linux-androideabi
```
Comment (La0) [inline nix/gecko_env.nix]: Does this options looks OK to you @marco-c  & @abpostelnicu  ?
This will not *only* build the android version, right ?

File: src/staticanalysis/bot/static_analysis_bot/__init__.py [hunk 4]
```diff
@@ -15,6 +15,7 @@
 CLANG_TIDY = 'clang-tidy'
 CLANG_FORMAT = 'clang-format'
 MOZLINT = 'mozlint'
+INFER = 'infer'
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/__init__.py]: :+1:

File: src/staticanalysis/bot/static_analysis_bot/infer/__init__.py [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+        {'credentials': {'clientId': os.getenv('TASKCLUSTER_CLIENT_ID'),
+                         'accessToken': os.getenv('TASKCLUSTER_ACCESS_TOKEN')}}
+    )
+    if job_name == 'linux64-infer':
+        job_names = ['linux64-infer', 'linux64-android-sdk-linux-repack',
+                     'linux64-android-ndk-linux-repack']
+        artifacts = ['public/build/infer.tar.xz',
+                     'project/gecko/android-sdk/android-sdk-linux.tar.xz',
+                     'project/gecko/android-ndk/android-ndk.tar.xz']
+        for i in range(len(job_names)):
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/__init__.py]: I would prefer:

```python
for job, artifact in zip(job_names, artifacts)
```

Or using a dict or even a list of tuples directly

File: src/staticanalysis/bot/static_analysis_bot/infer/__init__.py [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+            namespace = NAMESPACE.format(job_names[i], revision)
+            artifact_url = index.buildSignedUrl('findArtifactFromTask',
+                                                indexPath=namespace,
+                                                name=artifacts[i])
+            target = os.path.join(
+                os.environ['MOZBUILD_STATE_PATH'],
+                os.path.basename(artifacts[i]).split('.')[0],
+            )
+            cli_common.utils.retry(lambda: download(artifact_url, target))
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/__init__.py]: You should check at the end that the needed files are in the right place, with some asserts

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py [hunk 7]
```diff
@@ -0,0 +1,196 @@
...
+    '''
+    Infer runner
+    '''
+    def __init__(self, validate_checks=True):
+        self.binary = os.path.join(
+            os.environ['MOZBUILD_STATE_PATH'],
+            'infer', 'infer', 'bin', 'infer',
+        )
+        assert os.path.exists(self.binary), \
+            'Missing infer in {}'.format(self.binary)
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: :+1:

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py [hunk 8]
```diff
@@ -0,0 +1,196 @@
...
+        '''
+        assert isinstance(revision, Revision)
+        self.revision = revision
+
+        # Run all files in a single command
+        # through mach static-analysis
+        cmd = [
+            'gecko-env',
+            './mach', '--log-no-times', 'static-analysis', 'check-java'
+        ] + list(revision.files)
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: Maybe limit to java related files here, we had some issues with c/c++ and clang-tidy

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py [hunk 9]
```diff
@@ -0,0 +1,196 @@
...
+        logger.info('Running static-analysis', cmd=' '.join(cmd))
+
+        # Run command
+        try:
+            infer_output = subprocess.check_output(cmd, cwd=settings.repo_dir)
+        except subprocess.CalledProcessError as e:
+            logger.error('Mach static analysis failed: {}'.format(e.output))
+            raise
+
+        import json
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: All `import` on top of the file.

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py [hunk 10]
```diff
@@ -0,0 +1,196 @@
...
+        report_file = os.path.join(settings.repo_dir, 'infer-out', 'report.json')
+        infer_output = json.load(open(report_file))
+        # Dump raw clang-tidy output as a Taskcluster artifact (for debugging)
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: Please add a newline before a comment. That's just some nitpicking, but i find it way more readable

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py [hunk 11]
```diff
@@ -0,0 +1,196 @@
...
+        infer_output_path = os.path.join(
+            settings.taskcluster_results_dir,
+            '{}-infer.txt'.format(repr(revision)),
+        )
+        with open(infer_output_path, 'w') as f:
+            f.write(json.dumps(infer_output, indent=2))
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: I'd like to see a sample json output if you have some

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py [hunk 12]
```diff
@@ -0,0 +1,196 @@
...
+        '''
+        Parse infer output into structured issues
+        '''
+        if not infer_output:
+            logger.info('Infer could not generate any output.')
+            return []
+        # check if json is empty
+        nb_warnings = len(infer_output)
+        if not nb_warnings:
+            logger.info('Mach static analysis did not find any issue')
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: Specify that this is infer and/or java related here

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py [hunk 13]
```diff
@@ -0,0 +1,196 @@
...
+            return []
+        logger.info('Mach static analysis found some issues', nb=nb_warnings)
+        return [InferIssue(issue, revision) for issue in infer_output]
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: :100:

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py [hunk 14]
```diff
@@ -0,0 +1,196 @@
...
+        }
+
+    def is_problem(self):
+        return self.kind in ('ERROR')
+
+    def validates(self):
+        '''
+        Publish infer issues all the time
+        '''
+        return True
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: Are you sure about this ? Do we want to publish every issue on Phabricator (that's what will happend on prod here)? 

Is there some formatting only warnings ? We may want to skip those for now...

File: src/staticanalysis/bot/static_analysis_bot/revisions.py [hunk 15]
```diff
@@ -95,6 +107,7 @@ def __init__(self, description, api):
         assert isinstance(api, PhabricatorAPI)
         self.api = api
 
+        print(description)
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/revisions.py]: You might want to remove that :smiley:

File: src/staticanalysis/bot/static_analysis_bot/workflow.py [hunk 16]
```diff
@@ -146,8 +149,16 @@ def run(self, revision):
             revision.analyze_patch()
 
         with stats.api.timer('runtime.mach'):
-            # Only run mach if revision has any C/C++ files
-            if revision.has_clang_files:
+            # Only run mach if revision has any C/C++ or Java files
+            if revision.has_clang_files or revision.has_infer_files:
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/workflow.py]: I guess `./mach configure` is needed for infer too ? :cry:

File: src/staticanalysis/bot/static_analysis_bot/workflow.py [hunk 17]
```diff
@@ -165,18 +176,19 @@ def run(self, revision):
...
+                    else:
+                        logger.info('Skip clang-tidy')
+                    if CLANG_FORMAT in self.analyzers:
+                        analyzers.append(ClangFormat)
+                    else:
+                        logger.info('Skip clang-format')
 
             else:
-                logger.info('No clang files detected, skipping mach and clang-*')
+                logger.info('No clang or java files detected, skipping mach, infer and clang-*')
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/workflow.py]: Nice !

File: nix/gecko_env.nix (hunk 3 above)
Comment (sylvestre) [reply nix/gecko_env.nix]: yes, this is ok

File: .gitignore (hunk 1 above)
Comment (ReapOmen) [reply .gitignore]: `emacs` creates a lot of files which end in `~` so that it is possible to recover files. I will add a comment.

File: nix/gecko_env.nix (hunk 3 above)
Comment (La0) [reply nix/gecko_env.nix]: Thanks boss.

File: nix/gecko_env.nix (hunk 2 above)
Comment (ReapOmen) [reply nix/gecko_env.nix]: The environment variable does not change anything, it exists in order to increase the readability of the next two lines.

File: src/staticanalysis/bot/static_analysis_bot/infer/__init__.py (hunk 5 above)
Comment (ReapOmen) [reply src/staticanalysis/bot/static_analysis_bot/infer/__init__.py]: Yes, that makes more sense.

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py (hunk 8 above)
Comment (ReapOmen) [reply src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: The `check-java` command filters out all non-java files. Should I do this in the bot as well?

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py (hunk 14 above)
Comment (ReapOmen) [reply src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: `infer` does not look for formatting issues, only logical ones.

File: nix/gecko_env.nix (hunk 3 above)
Comment (abpostelnicu) [reply nix/gecko_env.nix]: Why do we specify the mobile application if we don't know yet that we have java files? This doesn't seem right, we are configuring an Android build and when we will generate the compile_commands.json we will have declarations related for Android so hence we will run clang-tidy for an Android build.
It would be nice to avoid this.

File: src/staticanalysis/bot/static_analysis_bot/workflow.py (hunk 16 above)
Comment (ReapOmen) [reply src/staticanalysis/bot/static_analysis_bot/workflow.py]: Yes

Comment (abpostelnicu) [reply src/staticanalysis/bot/static_analysis_bot/workflow.py]: Unfortunately yes, we need to generate the grade solution.

File: nix/gecko_env.nix (hunk 2 above)
Comment (La0) [reply nix/gecko_env.nix]: Don't expose it then, `gecko-env` is fickle, and i'd like to introduce as little changes as possible there

File: nix/gecko_env.nix (hunk 3 above)
Comment (ReapOmen) [reply nix/gecko_env.nix]: I have a workaround for this, but it will involve running `configure` and `export` two times.

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py (hunk 8 above)
Comment (La0) [reply src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: No it's ok then

File: nix/gecko_env.nix (hunk 3 above)
Comment (ReapOmen) [reply nix/gecko_env.nix]: I was thinking of having two separate mozconfigs. One that is generated by nix, and another one which I generate from within python.

File: src/staticanalysis/bot/static_analysis_bot/infer/infer.py (hunk 11 above)
Comment (ReapOmen) [reply src/staticanalysis/bot/static_analysis_bot/infer/infer.py]: https://gist.github.com/ReapOmen/77cb29bb08e30b721fddbd15dbbdc6a4

File: .gitignore [hunk 18]
```diff
@@ -26,3 +26,6 @@
 /src/**/yarn-error.log
 /tmp
 /w
+
+# ignore emacs generated files
```
Comment (sylvestre) [inline .gitignore]: Please remove that, this should be done in a separate PR

Comment (La0) [reply .gitignore]: Or in a [local system .gitignore](https://help.github.com/articles/ignoring-files/#create-a-global-gitignore)

File: src/staticanalysis/bot/static_analysis_bot/workflow.py [hunk 19]
```diff
@@ -181,26 +180,18 @@ def run(self, revision):
...
-                    analyzers.append(Infer)
-                    logger.info('Setup Taskcluster infer build...')
-                    setup_infer()
-
-                    # Mach pre-setup with mozconfig
-                    logger.info('Mach configure for infer...')
-                    run_check(['gecko-env', './mach', 'configure'], cwd=settings.repo_dir)
-                else:
-                    logger.info('Skip infer')
+                with AndroidConfig(settings):
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/workflow.py]: Use `with AndroidConfig` only when `INFER` is available

File: src/staticanalysis/bot/static_analysis_bot/infer/__init__.py [hunk 20]
```diff
@@ -63,3 +65,25 @@ def download(artifact_url, target):
     # Extract archive into destination
     with tarfile.open(fileobj=io.BytesIO(resp.content)) as tar:
         tar.extractall(target)
+
+
+class AndroidConfig():
```
Comment (La0) [inline src/staticanalysis/bot/static_analysis_bot/infer/__init__.py]: :+1:
